In [1]:
import tskit
import tszip
import numpy as np
import pandas as pd
import sc2ts

In [2]:
ds = sc2ts.Dataset("../data/viridian_mafft_2024-10-14_v1.vcz.zip")

In [4]:
ts = tszip.load("all_viridian.202409.trees.tsz")
ts

In [5]:
%%time
df_nodes = sc2ts.node_data(ts)

CPU times: user 6.98 s, sys: 547 ms, total: 7.52 s
Wall time: 7.52 s


In [6]:
%%time
df_muts = sc2ts.mutation_data(ts)

CPU times: user 1.1 s, sys: 167 ms, total: 1.27 s
Wall time: 1.27 s


In [7]:
df_muts

mutation_id  position  parent     node inherited_state derived_state  \
0                  0       267      -1  1009669               T             A   
1                  1       267      -1  4408344               T             G   
2                  2       269      -1  4681395               G             A   
3                  3       269      -1  2092107               G             A   
4                  4       269      -1   133742               G             T   
...              ...       ...     ...      ...             ...           ...   
3330802      3330802     29671      -1  4904110               A             G   
3330803      3330803     29671      -1  4900620               A             T   
3330804      3330804     29671      -1  3979258               A             T   
3330805      3330805     29671      -1  3988238               A             T   
3330806      3330806     29671      -1  4884543               A             T   

         num_descendants  num_inheritors       date  
0                      1               1 2021-07-30  
1                      1               1 2022-10-25  
2                      2               2 2020-04-05  
3                      1               1 2020-04-18  
4                      1               1 2020-05-08  
...                  ...             ...        ...  
3330802                2               2 2023-02-11  
3330803                2               2 2023-10-12  
3330804                1               1 2023-10-24  
3330805                1               1 2023-11-08  
3330806                3               3 2023-12-10  

[3330807 rows x 9 columns]

In [8]:
df_nodes

pango   sample_id  node_id  is_sample  is_recombinant  num_mutations  \
0           B  ERR4085584        0       True           False              0   
1           B  ERR4086315        1       True           False              0   
2           B  ERR4091746        2       True           False              0   
3           B  ERR4165155        3       True           False              0   
4           B  ERR4204267        4       True           False              0   
...       ...         ...      ...        ...             ...            ...   
5312038         node_1696  5312038      False           False              1   
5312039         node_1695  5312039      False           False              1   
5312040         node_1694  5312040      False           False              0   
5312041         node_1693  5312041      False           False              1   
5312042            node_1  5312042      False           False              0   

         max_descendant_samples       date  
0                             1 2020-04-09  
1                             1 2020-04-03  
2                             1 2020-04-03  
3                             1 2020-02-26  
4                             1 2020-03-04  
...                         ...        ...  
5312038                 4436338 2019-11-26  
5312039                 4436358 2019-11-25  
5312040                 4436366 2019-11-24  
5312041                 4436400 2019-11-23  
5312042                 4446273 2019-11-22  

[5312043 rows x 8 columns]

In [9]:
df_samples = df_nodes[df_nodes.is_sample].set_index("sample_id")

In [10]:
ds.metadata.field_descriptors()

field   dtype  \
0       Artic_primer_version  object   
1            Collection_date  object   
2                    Country  object   
3                  Date_tree  object   
4            Date_tree_order  object   
5                 Experiment  object   
6              First_created  object   
7                  Genbank_N   int16   
8          Genbank_accession  object   
9         Genbank_other_runs  object   
10          Genbank_pangolin  object   
11           Genbank_scorpio  object   
12         Genbank_tree_name  object   
13          In_Viridian_tree    bool   
14           In_intersection    bool   
15      In_may_2024_preprint    bool   
16                  Platform  object   
17                    Region  object   
18                 Run_count    int8   
19                    Sample  object   
20                     Study  object   
21                Viridian_N   int16   
22  Viridian_amplicon_scheme  object   
23         Viridian_cons_het    int8   
24         Viridian_cons_len   int16   
25         Viridian_pangolin  object   
26    Viridian_pangolin_1.29  object   
27           Viridian_result  object   
28          Viridian_scorpio  object   
29     Viridian_scorpio_1.29  object   

                                          description  
0   If known, the ARTIC primer scheme from ENA met...  
1                   collection_date from ENA metadata  
2   Country from ENA metadata. If ':' was in the e...  
3   A consensus date, using up 3 sources of data f...  
4   This helped define the order in which the samp...  
5              experiment_accession from ENA metadata  
6                 first_created from the ENA metadata  
7   Number of Ns in the GenBank consensus sequence...  
8       This is the GenBank accession of the assembly  
9   Any other run accessions associated with the G...  
10  Pangolin call from the GenBank consensus seque...  
11  Scorpio call from the GenBank consensus sequen...  
12             Name of the sample in the GenBank tree  
13  'T' or 'F' to indicate if the sample is in the...  
14  'T' or 'F' to indicate if the sample is in the...  
15  Whether or not this run was in the May 2024 pr...  
16              instrument_platform from ENA metadata  
17  Region from ENA metadata, if present. If ':' w...  
18       The number of runs from the sample accession  
19                 sample_accession from ENA metadata  
20                  study_accession from ENA metadata  
21  Number of Ns in the viridian consensus sequenc...  
22                 Amplicon scheme called by Viridian  
23                                                     
24  Length of the Viridian consensus sequence Viri...  
25  Pangolin call from the Viridian consensus sequ...  
26  Pangolin call from the Viridian consensus sequ...  
27  This is 'PASS' if Viridian finished and made a...  
28  Scorpio call from the Viridian consensus seque...  
29  Scorpio call from the Viridian consensus seque...

In [11]:
# Add fields in to verify Date and Pango validity.
df_samples = df_samples.join(ds.metadata.as_dataframe(["Date_tree", "Viridian_pangolin_1.29"]))
df_samples

pango  node_id  is_sample  is_recombinant  num_mutations  \
sample_id                                                                  
ERR4085584          B        0       True           False              0   
ERR4086315          B        1       True           False              0   
ERR4091746          B        2       True           False              0   
ERR4165155          B        3       True           False              0   
ERR4204267          B        4       True           False              0   
...               ...      ...        ...             ...            ...   
SRR20933483  BA.5.2.1  4446268       True           False              1   
SRR21370322  BA.5.2.1  4446269       True           False              1   
SRR21886505  BA.5.2.1  4446270       True           False              5   
SRR21635465  BA.5.2.1  4446271       True           False              0   
SRR21795271  BA.5.2.1  4446272       True           False              0   

             max_descendant_samples       date   Date_tree  \
sample_id                                                    
ERR4085584                        1 2020-04-09  2020-04-09   
ERR4086315                        1 2020-04-03  2020-04-03   
ERR4091746                        1 2020-04-03  2020-04-03   
ERR4165155                        1 2020-02-26  2020-02-26   
ERR4204267                        1 2020-03-04  2020-03-04   
...                             ...        ...         ...   
SRR20933483                       1 2022-07-11  2022-07-11   
SRR21370322                       1 2022-08-14  2022-08-14   
SRR21886505                       1 2022-09-21  2022-09-21   
SRR21635465                       1 2022-09-06  2022-09-06   
SRR21795271                       1 2022-09-12  2022-09-12   

            Viridian_pangolin_1.29  
sample_id                           
ERR4085584                       B  
ERR4086315                       B  
ERR4091746                       B  
ERR4165155                       B  
ERR4204267                       B  
...                            ...  
SRR20933483               BA.5.2.1  
SRR21370322               BA.5.2.1  
SRR21886505               BA.5.2.1  
SRR21635465               BA.5.2.1  
SRR21795271               BA.5.2.1  

[4446273 rows x 9 columns]

# Check metadata round trip

Check that we've successfully roundtripped the metadata from the MAT through tskit and via minimisation. We have used the Viridian_pangolin_1.29 field, which should be identical for all samples. The sample dates should be identica to the Date_tree field
also

In [12]:
np.all(df_samples["pango"] == df_samples["Viridian_pangolin_1.29"])

np.True_

In [13]:
np.all(df_samples["date"] == df_samples["Date_tree"])

np.True_

## Check genotype encoding

In [17]:
assert np.all(ts.samples() == df_samples.node_id)

In [45]:
ds_variants = ds.variants(sample_id=df_samples.index.values, position=ts.sites_position.astype(int))
ts_variants = ts.variants(alleles=tuple("ACGT"))
from tqdm.notebook import tqdm

per_site_diffs = []
for ds_var, ts_var in tqdm(zip(ds_variants, ts_variants), total=ts.num_sites):
    assert ts_var.alleles == tuple(ds_var.alleles[:4])
    missing = ds_var.genotypes == -1
    ambiguous = ds_var.genotypes > 3
    confident_calls = ~(missing | ambiguous)
    true_calls = ds_var.genotypes[confident_calls]
    stored = ts_var.genotypes[confident_calls]
    per_site_diffs.append(np.sum(true_calls != stored))

per_site_diffs = np.array(per_site_diffs)

  0%|          | 0/27499 [00:00<?, ?it/s]

In [49]:
non_zero = np.where(per_site_diffs != 0)
non_zero

(array([ 3988,  6072,  6074,  9382, 14339, 19901, 19916, 19919, 19920,
        19922, 19925, 19926, 19930, 19935, 19937, 19938, 20143, 20486,
        20488, 20489, 20491, 20493, 20495, 20944, 21051, 21077, 21137,
        22042, 22135, 23111, 24539, 24585, 25353, 25467, 25598, 25710,
        26041, 26807, 27222]),)

In [52]:
non_zero_pos = ts.sites_position[non_zero].astype(int)
non_zero_pos

array([ 4321,  6513,  6515, 10029, 15510, 21595, 21610, 21615, 21617,
       21621, 21624, 21627, 21632, 21637, 21639, 21641, 21846, 22195,
       22197, 22198, 22200, 22202, 22204, 22674, 22786, 22813, 22882,
       23854, 23948, 25000, 26530, 26577, 27384, 27507, 27638, 27752,
       28095, 28916, 29362])

In [53]:
non_zero_pos.shape

(39,)

In [54]:
ts.num_sites - non_zero_pos.shape[0]

27460

We have exact matches for 27460 of the sites in the data for which we have nucleotide calls.

In [72]:
import collections
data = []
for var in  ds.variants(position=non_zero_pos):
    alleles = np.append(var.alleles, ["N"])
    counter = collections.Counter(alleles[var.genotypes])
    data.append({"position": int(var.position), **{a:counter[a] for a in alleles}})


In [73]:
df_composition = pd.DataFrame(data, dtype=int)
df_composition

position        A        C        G        T        -      R     Y      S  \
0       4321      555  3347068        5  1115563        4      0  2229      1   
1       6513     5019        2  3656397      442   772494      9     0      0   
2       6515     5131       57        4  3656612   772571      0     6      0   
3      10029        0   807092        0  3660160        0      0   111      0   
4      15510        3     3167        1  4130426       17      0  1094      0   
5      21595       42  4269963      181   165995       94      0   791      3   
6      21610        8     7007       20  4426897      331      0    17      0   
7      21615        8     7731      958  4424251      250      0    15      0   
8      21617  4424814     7621      208       25      257     20     0      0   
9      21621    10856  4408687       21    10786     1241      0   174      7   
10     21624     7091    26345  4392633     4181     1154     15     1     77   
11     21627      291  4416767        3    13641      918      0   213      1   
12     21632       14     6723       56  4424443      549      0    31      0   
13     21637       53  3313015        9    13269  1104711      0   114      2   
14     21639     6721  3315800       29     2317  1104695      0   126      4   
15     21641      214       16  3314975    10495  1104655     23     0      5   
16     21846      224  2361710       95  1640456       43      0   946      6   
17     22195       70      474    58135  3549045   703845      0    32      0   
18     22197       74     2442    57876  4249147      526      0    31      0   
19     22198  4249665    57736      339       34      485      9     0      0   
20     22200   122845     1074   984275  3198823      420     35    10      0   
21     22202    57601  4248400      151      217      414      1    20      0   
22     22204      670    58082        7  4246687      402      0    18      0   
23     22674      149  2557988        0  1731782       32      0  1227      0   
24     22786  3155994  1007824      827      463       14     18    16      1   
25     22813       60      124  2556032  1530103        4      5    12      3   
26     22882      843      272  1464971  2522407        7      3    23      0   
27     23854  1740756  2550088       86     2527       10      4   101      1   
28     23948      105     2144  2569953  1881771       13      5    16     16   
29     25000       74  2565003       57  1912083        2      0  1274      3   
30     26530  3663322       27   746532      116      108  18072     0      0   
31     26577       58  2566841  1848371        3      123      4     4  20056   
32     27384       17   807411       19  3624015     1051      0   948      1   
33     27507  4321318   112428      363      161      885     12     2      0   
34     27638       33  1705336        4  2488061     7346      0    82      0   
35     27752       23  2508808       44  1748334     2493      0   151      2   
36     28095  4292686       28      935   105061     6355     22     1      0   
37     28916      820        9  2723179  1728031     2796     44     2      0   
38     29362       27  4422174        4    16090     3881      0  1581      0   

       W      K      M  B  D  H  V  .       N  
0      0      1      8  0  0  0  0  0   18723  
1      0     14      0  0  0  0  0  0   49780  
2      0      0      0  0  0  0  0  0   49776  
3      0      0      0  0  0  0  0  0   16794  
4      1      1      0  0  0  0  0  0  349447  
5      0      0      4  0  0  0  0  0   47084  
6      1      0      0  0  0  0  0  0   49876  
7      4      5      0  0  0  0  0  0   50935  
8      4      0     54  0  0  0  0  0   51154  
9      1      0     23  0  0  0  0  0   52361  
10     0    113      0  0  0  0  0  0   52547  
11     0      0     18  0  0  0  0  0   52305  
12     4      6      0  0  0  0  0  0   52331  
13     0      0     11  0  0  0  0  0   52973  
14     0      1      8  0  0  0  0  0   544

In [77]:
np.sum(df_composition["-"] > 500_000)

np.int64(6)

Of the 39 sites where we don't exactly reproduce the confident calls in the alignment, we have 6 sites that have more than 500,000 samples called as a gap character. It seems likely these are driven by differences in the alignment or perhaps or other factors.